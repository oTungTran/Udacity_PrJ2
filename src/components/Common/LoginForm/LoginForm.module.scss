.login-form {
    display: flex;
    flex-direction: column;
    background-color: transparent;
    align-items: center;
    justify-content: center;
    user-select: none;

    .wrapper {
        min-width: 40rem;

        h1 {
            text-align: center;
            color: #000;
        }

        .control {
            margin-bottom: 2rem;

            button {
                width: 100%;

            }
        }
    }
}