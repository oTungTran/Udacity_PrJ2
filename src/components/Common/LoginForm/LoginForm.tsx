import { FormEvent, useState } from "react";
import styles from "./LoginForm.module.scss";
import Input from "../Input/Input";
import Button from "../Button/Button";

export type LoginType = {
    username: string,
    password: string
};

export default function LoginForm({ submit, error }: { submit: (_: LoginType) => void, error?: string }) {

    const [formValue, setFormValue] = useState<LoginType>({
        username: '',
        password: ''
    });

    const changeFrm = (name: string, val: string) => {
        if (!name) {
            return;
        }
        setFormValue({ ...formValue, [name]: val });
    }

    const [errors, setErrors] = useState<LoginType | null>(null);

    const validateForm = () => {
        let newErrors: any = {};

        if (!formValue.username.trim()) {
            newErrors = { username: 'Username is required' };
        }

        if (!formValue.password.trim()) {
            newErrors.password = 'Password is required';
        }

        setErrors(newErrors);
        return Object.values(newErrors).length === 0;
    };

    const handleSubmit = (e: FormEvent) => {
        e.preventDefault();
        if (!validateForm()) {
            return;
        }
        submit(formValue);
    }

    return (
        <div className={styles['login-form']}>
            <div className={styles['wrapper']}>
                <h1 className={styles.h1}>Login</h1>
                <form className={styles.form} noValidate autoComplete="false" onSubmit={handleSubmit}>
                    <div className={styles.control}>
                        <Input
                            label="Username"
                            value={formValue.username}
                            name="username"
                            message={errors?.username || error}
                            onChange={e => changeFrm(e.target.name, e.target.value)} />
                    </div>
                    <div className={styles.control}>
                        <Input
                            label="Password"
                            name="password"
                            value={formValue.password}
                            message={errors?.password}
                            onChange={e => changeFrm(e.target.name, e.target.value)} />
                    </div>
                    <div className={styles.control}>
                        <Button
                            size="l"
                            className={styles.loginBtn}
                            onClick={handleSubmit}
                        >
                            Log in
                        </Button>
                    </div>
                </form>
            </div>
        </div>
    );
}