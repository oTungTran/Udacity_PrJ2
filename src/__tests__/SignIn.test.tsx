import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import LoginForm, { LoginType } from '../components/Common/LoginForm/LoginForm';

describe("<LoginForm />", () => {
    test('renders login form', async () => {
        const handleSubmit = jest.fn((e: LoginType) => { });

        render(<LoginForm submit={handleSubmit} />);

        const inpUsername = await screen.findByLabelText('Username');
        expect(inpUsername).toBeInTheDocument();

        const inpPassword = await screen.findByLabelText('Password');
        expect(inpPassword).toBeInTheDocument();

        const btnSubmit = screen.getByRole('button', { name: /Log in/i });
        expect(btnSubmit).toBeInTheDocument();
    });

    test('input username in form', async () => {
        const handleSubmit = jest.fn((e: LoginType) => { });

        render(<LoginForm submit={handleSubmit} />);

        const inpUsername = await screen.findByLabelText('Username');
        fireEvent.change(inpUsername, { target: { value: 'test' } });

        expect(inpUsername.getAttribute('value')).toBe("test");
    });

    test('login form fail', async () => {
        const handleSubmit = jest.fn((e: LoginType) => { });

        render(<LoginForm submit={handleSubmit} />);

        const inpPassword = await screen.findByLabelText('Password');
        fireEvent.change(inpPassword, { target: { value: 'abc321' } });

        const btnSubmit = screen.getByRole('button', { name: /Log in/i });
        fireEvent.click(btnSubmit);

        const usernameError = await screen.findByText(/Username is required/i);
        expect(usernameError).toBeInTheDocument();
    });

    test('login form success', async () => {
        let data: LoginType | undefined = undefined;
        const handleSubmit = jest.fn((e: LoginType) => (data = e));

        render(<LoginForm submit={handleSubmit} />);

        const inpUsername = await screen.findByLabelText('Username');
        fireEvent.change(inpUsername, { target: { value: 'tylermcginnis' } });

        const inpPassword = await screen.findByLabelText('Password');
        fireEvent.change(inpPassword, { target: { value: 'abc321' } });

        expect(data).toBeUndefined();

        const btnSubmit = screen.getByRole('button', { name: /Log in/i });
        fireEvent.click(btnSubmit);

        expect(handleSubmit).toHaveBeenCalled();
        expect(data).toMatchObject({ username: 'tylermcginnis', password: 'abc321' });
    });

    test("snapshot", () => {
        const handleSubmit = jest.fn((e: LoginType) => { });
        // eslint-disable-next-line testing-library/render-result-naming-convention
        const tree = render(<LoginForm submit={handleSubmit} />);
        expect(tree).toMatchSnapshot();
    });
});
