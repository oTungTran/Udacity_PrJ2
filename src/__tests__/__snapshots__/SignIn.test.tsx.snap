// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-form"
      >
        <div
          class="wrapper"
        >
          <h1
            class="h1"
          >
            Login
          </h1>
          <form
            autocomplete="false"
            class="form"
            novalidate=""
          >
            <div
              class="control"
            >
              <div
                class="wrapper"
              >
                <label
                  class="label"
                  for="username"
                >
                  Username
                </label>
                <div
                  class="inputBlock"
                >
                  <input
                    class="input "
                    id="username"
                    name="username"
                    placeholder="Username"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="control"
            >
              <div
                class="wrapper"
              >
                <label
                  class="label"
                  for="password"
                >
                  Password
                </label>
                <div
                  class="inputBlock"
                >
                  <input
                    class="input "
                    id="password"
                    name="password"
                    placeholder="Password"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="control"
            >
              <button
                class="loginBtn wrapper primary l  "
              >
                <span>
                  Log in
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-form"
    >
      <div
        class="wrapper"
      >
        <h1
          class="h1"
        >
          Login
        </h1>
        <form
          autocomplete="false"
          class="form"
          novalidate=""
        >
          <div
            class="control"
          >
            <div
              class="wrapper"
            >
              <label
                class="label"
                for="username"
              >
                Username
              </label>
              <div
                class="inputBlock"
              >
                <input
                  class="input "
                  id="username"
                  name="username"
                  placeholder="Username"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="control"
          >
            <div
              class="wrapper"
            >
              <label
                class="label"
                for="password"
              >
                Password
              </label>
              <div
                class="inputBlock"
              >
                <input
                  class="input "
                  id="password"
                  name="password"
                  placeholder="Password"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="control"
          >
            <button
              class="loginBtn wrapper primary l  "
            >
              <span>
                Log in
              </span>
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
